### ISSUE TITLE
Abstract classes

### ISSUE THREAD

**Issue by akshaysubr** [jameskermode/f90wrap#41]
Looks like abstract classes are not parsed properly. The code snippet

```
type, abstract :: shapes
    integer :: shapetype = 0
end type
```

is parsed as `type(abstract)` instead of an abstract `type(shapes)`.  However, since this is an abstract class, the `f90wrap_shapes__get__shapetype(this, shapetype)` subroutine would probably have to be handled differently.

**Comment by akshaysubr** [comment]
If the abstract class is not required in Python, looks like moving `f90wrap_shapes__get__shapetype(this, shapetype)` to classes that extend the abstract class is sufficient. Maybe a pre-processor that copies everything in the abstract class to it's child class would be the easiest solution.

**Comment by jameskermode** [comment]
That's correct, there's currently no support for any Fortran 2003 features. Mapping the abstract types to either standard Python classes or abstract base classes and then specifying the correct inheritance in the Python classes wrapped around the concrete Fortran derived types would be the best solution, and probably wouldn't be any more work than pre-processing.

**Comment by krystophny** [comment]
I took a quick look at the abstract types on https://github.com/itpplasma/f90wrap under `f90wrap/examples/issue41_abstract_classes`. It looked like one could get away with just ignoring the abstract base class at first, but then I hit two problems

1) Abstract types are still wrapped via pointers, leading to a compile error such as
```fortran
f90wrap_myclass_base.f90:12:24:

   12 |     allocate(this_ptr%p)
      |                        1
Error: Nonpolymorphic reference to abstract type at (1)
```

2) Factory functions or subroutines that allocate an instance do not work due to
```bash
DEBUG:f90wrap.transform:marking public symbol create_myclass
WARNING:f90wrap.transform:removing routine create_myclass due to allocatable/pointer arguments
```

Since I assume a meaningful abstract type support will also require factories, we are stuck at this point. @jameskermode do you have an idea how much effort implementation of allocatable output arguments would mean?

**Comment by jameskermode** [comment]
> @jameskermode do you have an idea how much effort implementation of allocatable output arguments would mean?

Probably not too bad. The reason it's not supported is because of the risk of memory leaks when memory is allocated in Fortran but would need to be freed from Python. One solution would be not to return the allocatable output argument directly, but instead wrap it in a Python class with a destructor that frees the Fortran-allocated memory when the Python object goes out of scope.

**Comment by krystophny** [comment]
Thanks for the fast reply! I opened another issue for that topic and will forward this to two colleagues who may be motivated to take a look at it.

**Comment by krystophny** [comment]
Status update: Since merging the fix for #227 I could play with factory methods. While types are working fine up to now, abstract types need polymorphic (`class`) variables when handling them. This introduces the following issues.

1. Extracting `deferred` method signatures from the abstract type may require a few (hopefully minor) changes in the parser.
2. Pointers to `class` variables cannot be set by an `allocatable` output of a function. This can be fixed by an intermediate variable that is then mapped to the pointer variable in the wrapper.
3. `class` variables are automatically deallocated when they go out of scope. With the `transfer` to an integer array also the information of the associated pointer is lost. This leads to early and/or double deallocation.

Point 3. is a bigger issue, since automatic finalisation and deallocation happens before the pointer is passed further to the Python interface.

As of now, I have played with a factory that only outputs one concrete implementation and changed the wrapper as follows. This does not yet resolve 3.

```fortran
subroutine f90wrap_myclass_factory__create_myclass(ret_myobject, impl_type)
    use myclass_factory, only: create_myclass
    use myclass_impl, only: myclass_impl_t
    implicit none

    class(myclass_impl_t), allocatable, target :: temp_obj

    type myclass_impl_t_ptr_type
        class(myclass_impl_t), pointer :: p
    end type myclass_impl_t_ptr_type
    type(myclass_impl_t_ptr_type) :: ret_myobject_ptr
    integer, intent(out), dimension(2) :: ret_myobject
    character*(*), intent(in) :: impl_type
    temp_obj = create_myclass(impl_type)
    ret_myobject_ptr%p => temp_obj
    ret_myobject = transfer(ret_myobject_ptr, ret_myobject)
end subroutine f90wrap_myclass_factory__create_myclass
```

**Comment by krystophny** [comment]
Seems like gfortran needs `dimension(4)` for `class` pointers by the way. Another way with another wrapper below has still the issue of automatic deallocation. Adding `save` to the `target` object works only for a single instance.

```fortran
subroutine f90wrap_myclass_factory__create_myclass(ret_myobject, impl_type)
    use myclass_factory, only: create_myclass
    use myclass_impl, only: myclass_impl_t
    implicit none

    type myclass_impl_wrapper_t
        class(myclass_impl_t), allocatable :: obj
    end type myclass_impl_wrapper_t

    type myclass_impl_t_ptr_type
        type(myclass_impl_wrapper_t), pointer :: p => NULL()
    end type myclass_impl_t_ptr_type

    type(myclass_impl_wrapper_t), target :: ret_myobject_wrapper
    type(myclass_impl_t_ptr_type) :: ret_myobject_ptr
    integer, intent(out), dimension(2) :: ret_myobject

    character*(*), intent(in) :: impl_type

    print *, "create_myclass: ", impl_type

    ret_myobject_wrapper%obj = create_myclass(impl_type)
    ret_myobject_ptr%p => ret_myobject_wrapper
    ret_myobject = transfer(ret_myobject_ptr, ret_myobject)
end subroutine f90wrap_myclass_factory__create_myclass
```

**Comment by krystophny** [comment]
Rebased now on #236 . Additional changes required:

- [ ] Parser: Handle `deferred` procedures in abstract type definitions and their bindings to according interfaces
- [ ] Code generation: Do not generate default constructor for abstract types

For clean separation, one would introduce additional attributes for nodes in the parse tree, e.g. `is_abstract`.

**Comment by krystophny** [comment]
- [ ] Modify wrapper for factories to assign correct subclass, or find some other solution (`fortrantype.py`). Right now it does

```python
def create_myclass(impl_type):
    """
    myobject = create_myclass(impl_type)
    
    
    Defined at myclass_factory.f90 lines 7-18
    
    Parameters
    ----------
    impl_type : str
    
    Returns
    -------
    myobject : Myclass_T
    
    """
    myobject = _itest.f90wrap_myclass_factory__create_myclass(impl_type=impl_type)
    myobject = f90wrap.runtime.lookup_class("itest.myclass_t").from_handle(myobject, \
        alloc=True)
    return myobject
```

**Comment by krystophny** [comment]
[ ] Use `ABC` in Python wrapper generator?
